alf/config_util_test.py

```diff
@@ -182,6 +182,63 @@ def sole_init_test_env(x):
         self.assertEqual(alf.get_config_value("sole_init_test_twice.x"), 1)
         self.assertEqual(alf.get_config_value("sole_init_test_env.x"), 1)
 
+        # Test override_config doesn't doesn't overwrite for immutable values.
+        @alf.configurable
+        def override_on_immutable(x):
+            pass
+
+        alf.config("override_on_immutable", x=0, mutable=False)
+        alf.override_config("override_on_immutable", x=1)
+        self.assertEqual(alf.get_config_value("override_on_immutable.x"), 0)
+
+        # Test override_config doesn't doesn't overwrite for immutable values.
+        @alf.configurable
+        def override_on_immutable_and_sole_init(x):
+            pass
+
+        alf.config(
+            "override_on_immutable_and_sole_init",
+            x=0,
+            sole_init=True,
+            mutable=False)
+        alf.override_config("override_on_immutable_and_sole_init", x=1)
+        self.assertEqual(
+            alf.get_config_value("override_on_immutable_and_sole_init.x"), 0)
+
+        # Test that pre_config calls work before a sole_init config call
+        @alf.configurable
+        def pre_config_before(x):
+            pass
+
+        alf.pre_config({"pre_config_before.x": 0})
+        alf.config("pre_config_before", x=1)
+        alf.override_config("pre_config_before", x=1)
```

Comment:
We use pre_config in deployment app to specify deploy mode etc..  In order to avoid them from being overridden, should we specify them as immutable?  Or should this override_config call throw error?

Reply:
`pre_config` already sets all values as `mutable=False`. This is how it and `--conf_params` prevent their config values from being overwritten.

Reply:
Ah, my mistake.